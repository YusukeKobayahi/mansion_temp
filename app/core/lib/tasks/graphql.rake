# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
