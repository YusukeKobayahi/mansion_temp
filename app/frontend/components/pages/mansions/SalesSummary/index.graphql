fragment MsLibraryMansionsSalesSummary on Mansion {
  constructedIn
  age
  name
  salesHistorySummaries {
    layout
    minPrice
    maxPrice
    minExclusiveArea
    maxExclusiveArea
  }
}
