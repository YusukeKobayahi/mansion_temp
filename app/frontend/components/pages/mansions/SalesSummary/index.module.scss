@import "~/styles/variables";
@import "~/styles/utils";

.container {
  max-width: 730px;
  margin: 20px auto 0;
  color: #746744;

  @include screenMd {
    padding: 0 16px;
  }
}

.content {
  width: 100%;
  padding: 30px;
  box-shadow: 2px 3px 0px #00000029;
  border: 1px solid #a5895c;
  border-radius: 10px;
  background-color: #fcfcfc;

  @include screenMd {
    padding: 26px 16px;
  }
}

.priceArea {
  &__title {
    font-size: 14px;

    @include screenMd {
      text-align: center;
    }
  }

  &__price {
    padding: 15px 0;
    border-bottom: 1px solid #b19c7a;
    font-size: 32px;
    font-weight: bold;
    text-align: center;

    @include screenMd {
      font-size: 20px;
    }
  }

  &__date {
    padding: 20px;
    text-align: center;
    font-size: 18px;
    font-weight: bold;

    @include screenMd {
      padding: 25px;
      font-size: 14px;
    }
  }
}

.summaryArea {
  &__title {
    font-size: 14px;

    @include screenMd {
      text-align: center;
    }
  }

  &__list {
    list-style-type: disc;
    margin-top: 15px;
    margin-left: 35px;
    line-height: 1.4;

    @include screenMd {
      margin-left: 16px;
    }
  }

  &__item {
    margin-top: 6px;
    font-size: 20px;

    &:first-child {
      margin-top: 0;
    }

    @include screenMd {
      font-size: 14px;
    }
  }

  &__emphasisText {
    font-weight: bold;
  }
}

.notes {
  color: $color-text-main;
  margin-top: 1.5em;
  font-size: 14px;

  li {
    line-height: 1.6;
    text-indent: -1em;
    padding-left: 1em;
  }

  @include screenMd {
    font-size: 12px;
  }
}
