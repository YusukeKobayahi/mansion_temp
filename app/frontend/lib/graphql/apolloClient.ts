// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
