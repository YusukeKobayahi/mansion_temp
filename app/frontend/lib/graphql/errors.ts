export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
