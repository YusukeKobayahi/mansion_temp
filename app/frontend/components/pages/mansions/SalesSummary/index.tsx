import { useDeepCompareMemo } from "use-deep-compare";
import SalesSummaryBuilder from "~/lib/mansions/SalesSummaryBuilder";
import styles from "./index.module.scss";
import { MsLibraryMansionsSalesSummaryFragment } from "~/graphql/generated";
import SidelineHeading from "~/components/commons/headings/SidelineHeading";

export interface SalesSummaryProps {
  data: MsLibraryMansionsSalesSummaryFragment;
}

const SalesSummary: React.FC<SalesSummaryProps> = ({
  data,
}: SalesSummaryProps) => {
  const builder = useDeepCompareMemo(
    () => new SalesSummaryBuilder(data.salesHistorySummaries),
    [data.salesHistorySummaries]
  );

  const layoutSummaries = useDeepCompareMemo(() => {
    return builder.buildLayoutSummaries();
  }, [data.salesHistorySummaries]);

  return (
    <div className={styles.container}>
      <h3>
        <SidelineHeading>{data.name}の相場価格</SidelineHeading>
      </h3>
      <div className={styles.content}>
        <section className={styles.summaryArea}>
          <h3 className={styles.summaryArea__title}>参考中古価格</h3>
          <ul className={styles.summaryArea__list}>
            {layoutSummaries.map((s, i) => (
              <li className={styles.summaryArea__item} key={i}>
                <span className={styles.summaryArea__emphasisText}>
                  {s.layout}
                </span>
                {` / ${s.info}`}
              </li>
            ))}
          </ul>
        </section>
      </div>
    </div>
  );
};

export default SalesSummary;
