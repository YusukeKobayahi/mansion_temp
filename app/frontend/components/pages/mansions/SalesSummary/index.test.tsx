import { render } from "@testing-library/react";
import mockData from "../fixtures/mockData.json";
import { MansionQuery } from "~/graphql/generated";
import SalesSummary from ".";

describe("components/pages/mansions/SalesSummary", () => {
  it("正常に表示される", () => {
    const mock = mockData as MansionQuery;
    const { asFragment } = render(<SalesSummary data={mock.mansion} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
