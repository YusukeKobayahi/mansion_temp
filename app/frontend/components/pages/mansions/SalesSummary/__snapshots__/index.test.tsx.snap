// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/mansions/SalesSummary 正常に表示される 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3>
      <div
        class="heading"
      >
        <span
          class="inner"
        >
          リヴシティ神田の相場価格
        </span>
      </div>
    </h3>
    <div
      class="content"
    >
      <section
        class="summaryArea"
      >
        <h3
          class="summaryArea__title"
        >
          参考中古価格
        </h3>
        <ul
          class="summaryArea__list"
        >
          <li
            class="summaryArea__item"
          >
            <span
              class="summaryArea__emphasisText"
            >
              1DK
            </span>
             / 2,700万円 ~ 4,780万円 / 45.16㎡
          </li>
          <li
            class="summaryArea__item"
          >
            <span
              class="summaryArea__emphasisText"
            >
              1K
            </span>
             / 2,000万円 ~ 3,200万円 / 20.72㎡ ~ 30.24㎡
          </li>
        </ul>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
